library/src/main/java/org/apache/apex/malhar/lib/wal/FSWindowDataManager.java

```diff
@@ -51,291 +66,607 @@
+      return null;
+    }
 
-            Path loperatorPath = new Path(appPath, Integer.toString(loperator));
-            if (fs.listStatus(loperatorPath).length == 0) {
-              //The operator was deleted and it has nothing to replay.
-              deletedOperators.remove(loperator);
-              fs.delete(loperatorPath, true);
-            }
-          } else if (loperator == operatorId) {
-            storageAgent.delete(loperator, lwindow);
-          }
+    FileSystemWAL.FileSystemWALReader reader = wal.getReader();
+
+    while (reader.getCurrentPointer() == null ||
+        reader.getCurrentPointer().compareTo(wal.walEndPointerAfterRecovery) < 0) {
+      long currentWindow;
+
+      if (wal.retrievedWindow == null) {
+        wal.retrievedWindow = readNext(reader);
+        Preconditions.checkNotNull(wal.retrievedWindow);
+      }
+      currentWindow = Longs.fromByteArray(wal.retrievedWindow.toByteArray());
+
+      if (windowId == currentWindow) {
+        Slice data = readNext(reader);
+        Preconditions.checkNotNull(data, "data is null");
+
+        wal.windowWalParts.put(currentWindow, reader.getCurrentPointer().getPartNum());
+
+        Slice nextSlice = readNext(reader); //null, deleted or next window
+
+        if (nextSlice == null) {
+          // the current window is not deleted
+          wal.retrievedWindow = null;
+          return fromSlice(data);
+        } else if (nextSlice.equals(DELETED)) {
+          //current window is deleted;
+          wal.retrievedWindow = null;
+          //window end is after deleted
+          return null;
+        } else {
+          //next window
+          wal.retrievedWindow = nextSlice;
+          return fromSlice(data);
+        }
+      } else if (windowId < currentWindow) {
+        //no artifact saved corresponding to that window and artifact is not read.
+        return null;
+      } else {
+        //windowId > current window so we skip the data
+        skipNext(reader);
+        wal.windowWalParts.put(currentWindow, reader.getCurrentPointer().getPartNum());
+
+        Slice nextSlice = readNext(reader); //null, deleted or next window
+
+        if (nextSlice == null) {
+          //nothing else to read
+          wal.retrievedWindow = null;
+          return null;
+        } else if (nextSlice.equals(DELETED)) {
+          wal.retrievedWindow = null;
+        } else {
+          wal.retrievedWindow = nextSlice;
         }
-        iterator.remove();
       }
     }
+    return null;
+  }
 
-    if (fs.listStatus(new Path(appPath, Integer.toString(operatorId))).length > 0) {
-      long[] windowsAfterReplay = storageAgent.getWindowIds(operatorId);
-      Arrays.sort(windowsAfterReplay);
-      for (long lwindow : windowsAfterReplay) {
-        if (lwindow <= windowId) {
-          storageAgent.delete(operatorId, lwindow);
-        }
+  public Map<Long, Object> retrieveAllWindows() throws IOException
```

Comment:
Will move this to `IncrementalCheckpointManager`. `AbstractManagedState` uses it.